src/invisible/how-tos/data-collection-with-the-companion-app/achieve-super-precise-time-sync.md

```diff
@@ -0,0 +1,60 @@
+
+
+### Force Syncing to the Master Clock on Demand
+The easiest way to achieve more accurate time synchronization is to force a sync-up to the master clock of all devices before starting data collection. The drift error takes a while to accumulate and is negligible for at least several hours after a sync-up. While the error is <~20 ms directly after sync-up, it should still be `<~150 ms` after 24 hours.
+
+A sync-up can usually be forced by toggling automatic determination of the current time off and back on in the operating system's settings. In Android, for example, the `Date & Time` settings in the `System` settings have a toggle called `Use network-provided time`. Whenever this toggle is turned on, the system syncs up.
+
+Depending on the operating system, devices use different master clock servers to sync with. Ideally, all devices would sync to the exact same master clock to avoid small errors between different masters. Some operating systems allow to specify which server to use. Android is using the following ones depending on the region:
+
+```
+Asia
+NTP_SERVER=asia.pool.ntp.org
+
+Europe
+NTP_SERVER=europe.pool.ntp.org
+
+North America
+NTP_SERVER=north-america.pool.ntp.org
+```
+
+### Improving Synchronization further
+While an error of `<20 ms` is sufficient for most applications, some require even better synchronization. To improve further, you can estimate the exact offset between the clock used by Pupil Invisible and the external clock down to the millisecond.
+
+This can be done using the `TimeOffsetEstimator` of the real-time API. Using the following code, you can estimate the offset between the Pupil Invisible clock and the clock of the host executing the code.
+::: tip
+**Dependency**: `pip install "pupil-labs-realtime-api>=1.1.0"`
+:::
+```python
+from pupil_labs.realtime_api.simple import discover_one_device
+
+# Look for devices. Returns as soon as it has found the first device.
+print("Looking for the next best device...")
+device = discover_one_device(max_search_duration_seconds=10)
+if device is None:
+    raise SystemExit("No device found.")
+
+estimate = device.estimate_time_offset()
+if estimate is None:
+    device.close()
+    raise SystemExit("Pupil Companion app is too old")
+
+print(f"Mean time offset: {estimate.time_offset_ms.mean} ms")
+print(f"Mean roundtrip duration: {estimate.roundtrip_duration_ms.mean} ms")
+
+device.close()
+```
+
+Using continuous offset estimates like this, you can precisely compensate for clock drifts by correcting the respective timestamps.
+
+When using a wifi connection, the speed of data transfer can vary at times, leading to inaccuracies in the estimation. If precise synchronization down to a few milliseconds is required, it is recommended to use an ethernet connection. You can connect an ethernet cable to the phone using an appropriate USB-C adapter or hub.
```

Comment:
I feel like this paragraph should be moved to a more prominent location, since the whole super millisecond accuracy kind of depends on a wired connection, unless I'm misunderstanding.

I think a good location would be perhaps as a tip/aside after this sentence: 
"This can be done using the TimeOffsetEstimator of the real-time API."

Reply:
TBH I am not sure how large the influence of the wifi is. The `TimeOffsetEstimator` is running a series of measurements and averages across them, which should largely alleviate inaccuracies from noise in the wifi speed.
I *think* the accuracy just decreases from +-1 ms to +-3-4 ms max, but we might want to measure this to be sure.

Reply:
Ah ok. In that case, I think it's worth measuring it so the reader has a rough idea of the impact. Otherwise, the statement is a bit nebulous imo